### PR TITLE
fea/link ir to inference and fc fuse

### PR DESCRIPTION
**Pull request by Superjomn**
relay on https://github.com/PaddlePaddle/Paddle/pull/12783

### CODE REVIEW

File: paddle/fluid/framework/ir/fc_fuse_pass.cc [hunk 1]
```diff
@@ -0,0 +1,198 @@
...
+    GET_NODE(mul_weight);              // Y
+    GET_NODE(elementwise_add_tmpvar);  // bias
+    GET_NODE(elementwise_add_out);     // Out
+    GET_NODE(mul);                     // MUL op
+    GET_NODE(elementwise_add);         // ELEMENT_ADD op
+    GET_NODE(mul_out);                 // tmp
+#undef GET_NODE
+
+    // Create an FC Node.
+    // NOTE No FC op is valid currently, this is just a demo.
```
Comment (panyx0718) [inline paddle/fluid/framework/ir/fc_fuse_pass.cc]: Remove this line if it's available now?

File: paddle/fluid/framework/ir/fc_fuse_pass.cc [hunk 2]
```diff
@@ -0,0 +1,198 @@
...
+
+    g_descs.emplace_back();
```
Comment (panyx0718) [inline paddle/fluid/framework/ir/fc_fuse_pass.cc]: Do you want to clear g_descs at the very beginning of the pass or CHECK g_descs is empty?

I suggest use Graph->Set<>() to save it in graph.

File: paddle/fluid/framework/ir/fc_fuse_pass_tester.cc [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+namespace framework {
+namespace ir {
+
+void SetOp(ProgramDesc* prog, const std::string& type,
+           const std::vector<std::string>& inputs,
+           const std::vector<std::string>& outputs) {
+  auto* op = prog->MutableBlock(0)->AppendOp();
+  op->SetType(type);
+  op->SetInput("Xs", inputs);
+  op->SetOutput("Xs", outputs);
```
Comment (panyx0718) [inline paddle/fluid/framework/ir/fc_fuse_pass_tester.cc]: different output name?

File: paddle/fluid/framework/ir/graph.h [hunk 4]
```diff
@@ -134,7 +136,6 @@ class Graph {
     return ret;
   }
 
- private:
   // This method takes ownership of `node`.
   ir::Node *AddNode(ir::Node *node) {
```
Comment (panyx0718) [inline paddle/fluid/framework/ir/graph.h]: make it private if not used?

File: paddle/fluid/framework/ir/infer_clean_graph_pass.cc [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+class InferCleanGraphPass : public Pass {
+ public:
+  virtual ~InferCleanGraphPass() {}
+
+ protected:
+  std::unique_ptr<ir::Graph> ApplyImpl(std::unique_ptr<ir::Graph> graph) const {
+    PADDLE_ENFORCE(graph.get());
+
+    auto is_ctrol_dep_node = [](Node* x) {
+      return x && x->IsVar() && !x->Var();
```
Comment (panyx0718) [inline paddle/fluid/framework/ir/infer_clean_graph_pass.cc]: This is not always true. Some vars don't have VarDesc and not control dep node

There is an API called `IsControlDepVar`

File: paddle/fluid/framework/ir/node.h [hunk 6]
```diff
@@ -34,14 +34,16 @@ class Node {
...
-        var_desc_(var_desc),
+        var_desc_(new VarDesc(*var_desc->Proto())),
         op_desc_(nullptr),
         type_(Type::kVariable) {}
 
   explicit Node(OpDesc* op_desc)
       : name_(op_desc->Type()),
         var_desc_(nullptr),
-        op_desc_(op_desc),
+        op_desc_(
```
Comment (panyx0718) [inline paddle/fluid/framework/ir/node.h]: Is this change necessary? It's wrong to put nullptr BlockDesc

File: paddle/fluid/framework/ir/infer_clean_graph_pass.cc (hunk 5 above)
Comment (Superjomn) [reply paddle/fluid/framework/ir/infer_clean_graph_pass.cc]: this will make the inference abnormal.

**Review by panyx0718 — APPROVED** [review]
